File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
    vector<int> st;

        for (int ast : asteroids) {
            bool alive = true;

            while (alive && ast < 0 && !st.empty() && st.back() > 0) {
                if (abs(ast) > st.back()) {
                    st.pop_back();
                } else if (abs(ast) == st.back()) {
                    st.pop_back();
                    alive = false;
                } else {
                    alive = false;
                }
            }

            if (alive) {
                st.push_back(ast);
            }
        }

        return st;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.dart
class Solution {
  List<int> asteroidCollision(List<int> asteroids) {
    List<int> st = [];

    for (int ast in asteroids) {
      bool alive = true;

      while (alive && ast < 0 && st.isNotEmpty && st.last > 0) {
        if (ast.abs() > st.last) {
          st.removeLast();
        } else if (ast.abs() == st.last) {
          st.removeLast();
          alive = false;
        } else {
          alive = false;
        }
      }

      if (alive) {
        st.add(ast);
      }
    }

    return st;
  }
}